Train Text from scratch

xaedes opened the pull request:
I improved the training process (https://github.com/ggerganov/llama.cpp/pull/1360) by some orders of magnitudes, replacing naive backward passes with dedicated operations. 
The training can now also use flash attention to support bigger context sizes.

There is a self contained example which allows to train small llama models from scratch.
The vocabulary that is used will be loaded from a source llama model.
To be able to resume training from previous runs a training checkpoint file is used to store the optimizer context and model weights.
After training the loaded checkpoint can be exported as llama compatible model file in F32 format. 

List of all new operations:
- GGML_OP_REPEAT_BACK : To get rid of 10 operations per `repeat` backward pass.
- GGML_OP_OUT_PROD : Similar to MAT_MUL, but the second dimensions must be equal instead of the first dimension. Internally uses vec_mad instead of vec_dot which is used by matmul. 
- GGML_OP_SOFT_MAX_BACK : To get rid of big intermediate matrices, reducing runtime from quadratic to linear and eliminate quadratic memory overhead, now there is none.
- GGML_OP_FLASH_ATTN_BACK : Flash attention combines `softmax(K*Q*V)` into one operation. Using this with the new backward pass saves a lot of operations and memory overhead during training.
- GGML_OP_CROSS_ENTROPY_LOSS : Combines `-sum(log(softmax(a))*b)` into one dedicated operation. As the vocabulary is quite big (32k) the loss function deals with big matrices, as dedicated operation a lot of memory overhead can be saved.
- GGML_OP_CROSS_ENTROPY_LOSS_BACK : Backward pass of cross entropy loss.

Changes to optimizers:

AdamW was easy to implement by changing Adam. 
When the corresponding parameter is zero it mathematically simplifies to the regular Adam optimizer.
So instead of implementing a whole new optimizer, AdamW weight decay was implemented directly in Adam.

Adam tracks statistics about the gradients, which are very important for training. 
It was necessary to persist the state of the optimizer between ggml_opt calls.
For this I added `struct ggml_opt_context`, `ggml_opt_init()` and `ggml_opt_resume()`.
The regular `ggml_opt()` call will now internally create a new opt context, initialize it and then use resume to optimize using this fresh context.
This moves all the memory allocation done by the optimizers into ggml_opt_init.

I investigated the use of flash_ff, but it seems to implement a different feedforward than the one used in llama. 
Having a dedicated operation for the feedforward of llama could also save a lot of memory overhead, allowing bigger n_embd and n_ff.
Might be worth to look into.

There is still a lot of unnecessary memory overhead. 
In the llama eval function similar overhead is avoided using scratch buffers.
Would be good to find a way to do similar with the (automatically generated) backward pass.

Other noteworthy changes:
- added llama api function to get vocabulary data from llama context
- added a fix to llama_model_load_internal for models with n_layer<32 to be recognized as MODEL_7B, so we can load small self trained models - they are probably smaller than 7B.
- bugfix in llama_sample_token_mirostat_v2 when all candidates are filtered out, which can happen with freshly trained models

Also see:
- https://github.com/ggerganov/ggml/issues/8#issuecomment-1518465663
- https://github.com/ggerganov/llama.cpp/pull/1360

xaedes committed:
add python bindings for functions to get and set the whole llama state
(rng, logits, embedding and kv_cache)

xaedes committed:
fix decoding error. adds errors=ignore parameter

xaedes committed:
add python wrapper

https://gist.github.com/abetlen/2b90e5f153f6efd00931d098de5c73ce

xaedes committed:
update python bindings

xaedes committed:
add text generating baby-llama from scratch example

xaedes committed:
fix race condition bug in ggml_compute_forward_diag_mask_f32

xaedes committed:
implement ggml_soft_max_back for more performant backward pass of soft_max

avoids creating big intermediate matrices of size n_embd x n_embd for llama layers and n_vocab x n_vocab for cross entropy loss

xaedes committed:
improve softmax backward pass

go from quadratic runtime to linear runtime by simplifying the formulas

xaedes committed:
fix race condition bug in non-inplace ggml_compute_forward_diag_mask_f32

memcpy needs to be synchronized across threads to avoid race conditions.
=> do it in INIT phase

xaedes committed:
fix bug in ggml_compute_forward_soft_max_back_f32 on DEBUG build

xaedes committed:
improve performance of mul_mat backward pass

avoid transpose by using mul_mat with swapped arguments

xaedes committed:
avoid printing too much newlines in baby-llama-text

xaedes committed:
activate threading in baby-llama-text

xaedes committed:
add ggml_out_prod and use it for mul_mat backward pass for improved performance

performance stats report improvement from 37 seconds to 16 seconds runtime during my training tests

xaedes committed:
better weight initialization improves training convergence at start

xaedes committed:
better weight initialization improves training convergence at start

xaedes committed:
improve ggml_out_prod performance

- change iteration order (>15s -> 10s runtime)
- parallelize over one more dimension: over dst matrix rows (10s -> <5s runtime)

xaedes committed:
add llama sampler, shuffle samples and constrain sampling to tokens occurring in train data

xaedes committed:
fix get_samples call, add model tensor names, increase model size, start training samples after newline

xaedes committed:
save train trained model to checkpoint and load model to be trained from checkpoint

xaedes committed:
use inplace functions where possible

xaedes committed:
initialize rng with srand

xaedes committed:
use different arguments for input and output checkpoint

xaedes committed:
ggml fixes to support backward pass on inplace operations

xaedes committed:
remove duplicate include

xaedes committed:
fix cross entropy loss

- add target probabilities for each sample which is then used in cross entropy loss

xaedes committed:
print used memory before and after optimization

xaedes committed:
sample with non-greedy sampling parameters at the end of training

xaedes committed:
add cmake target for baby-llama-text

xaedes committed:
add ggml_add1_inplace to header

xaedes committed:
enable gradient propagation for inplace add1 and scale operations

those functions backward passes don't need the original src0, so they also work when forward is inplace

xaedes committed:
implement AdamW in ggml_opt_adam by adding weight decay parameter (default 0.001f)

also add a schedule parameter (default 1.0f) that can be used to scale alpha and decay according to learning schedule.
setting the decay parameter to zero disables AdamW resulting in normal Adam optimizer.

since the difference between Adam and AdamW is minimal it is not implemented as another optimizer, but integrated into the existing Adam optimizer.

xaedes committed:
use inplace operations in cross_entropy_loss

xaedes committed:
fix random weight initialization scale

xaedes committed:
add missing default parameters for adam optimizer

xaedes committed:
add ggml_opt_context, so that we can properly resume training

otherwise the optimizer states, tracking statistics about the error function and its derivates,
will reset to zero each time ggml_opt is called, hindering convergence on resumed training.

now the optimizer context and all its memory is stored in a separate struct.

xaedes committed:
fix bug in llama_sample_token_mirostat_v2

when all candidates are filtered out through mu threshold, the following soft_max operation will fail.
so keep at least one.

xaedes committed:
add forward function without using cache, for more performant training

during training on whole samples no cache is required.
removing the cache and simplifying the remaining code results in performance and memory usage improvement.

xaedes committed:
print suppressed newline tokens as string "\n"

printing too much actual newlines is suppressed to avoid flooding the console.

xaedes committed:
store optimizer state in training checkpoint and add learning schedule

persistent optimizer state allows to resume training without resetting the optimizer
learning schedule consists of linear warmup ramp followed by cosine decay with restarts

xaedes committed:
remove unused functions

xaedes committed:
fix bug in get_samples which corrupted training targets

xaedes committed:
save checkpoint only when it was trained

xaedes committed:
simplify code

xaedes committed:
remove trailing whitespace

xaedes committed:
simplify backward pass for SQRT

xaedes committed:
replace inefficient repeat backward pass with dedicated repeat_back operation

xaedes committed:
add ggml_cross_entropy_loss with backward pass for faster training

cross entropy loss can also be implemented using softmax and log, but as dedicated operation it is faster and especially avoids unnecessary memory overhead.

xaedes committed:
add tests for cross_entropy_loss backward pass

finite differences regularly results in estimated gradient of zero, despite the backward pass giving non zero gradient.
_probably_ the finite differences fails due to numerical issues

xaedes committed:
use ggml_cross_entropy_loss in text training example

xaedes committed:
remove trailing whitespace

xaedes committed:
slightly improve how cross entropy loss is compute

btw: directly implemented cross entropy loss seems to have way lower magnitudes than when implemented with softmax and log.
probably the input to log gets closer to zero due to float numerics.
maybe the multiplication by (1.0-eps)/sum is more accurate..

xaedes committed:
add llama_get_vocab to get the vocabulary as output parameters

xaedes committed:
set default model.type for unknown models with few layers

xaedes committed:
add export of training checkpoint to llama compatible model file

xaedes committed:
get vocabulary for exporting training checkpoint to llama compatible model file

ggerganov committed:
ci : disable temporary

xaedes committed:
implement backward pass of flash attention

xaedes committed:
bugfixes for backward pass of flash attention

xaedes committed:
test flash attention backward pass

need to set loose error bounds to pass.
the finitie differences are close to numeric limits and often return quite different values than the backward pass.
reducing eps further lets the gradients vanish completely.
likewise setting eps to big results in wronger values.
the softmax in the middle of the function is probably the most responsible for the numeric issues using finite differences.

xaedes committed:
add option to train with flash attention and move options to the top of the main function

training from scratch also works with flash attention
training convergence and generation results after fix number of iterations are worse than when not using flash attention.
maybe there still lingers a bug in the flash attention backward pass?
but training works, just with slower convergence.

flash attention is still worth to use, because it requires way less memory and is faster with high n_ctx

xaedes committed:
add train_params and command line option parser

xaedes committed:
remove unnecessary comments

xaedes committed:
add train params to specify memory size

xaedes committed:
Merge branch 'master' into text-from-scratch

# Conflicts:
#	ggml.c : number of operations and GGML_ASSERT vs assert

xaedes committed:
remove python bindings

xaedes committed:
rename baby-llama-text to train-text-from-scratch

Green-Sky commented on llama.cpp:
```diff
@@ -936,6 +936,12 @@ static void llama_model_load_internal(
+            default:
+                {
+                    if (hparams.n_layer < 32) {
+                        model.type = e_model::MODEL_7B;
```
we are adding the even smaller 3B (OpenLLaMa) here https://github.com/ggerganov/llama.cpp/pull/1588

klosax commented:
We may need the new [Sophia Optimizer](https://arxiv.org/abs/2305.14342) for a 2X increase in training speed compared to Adam.

mudler commented:
JFYI compiling locally fails with:

```
/home/mudler/_git/llama.cpp-train/examples/train-text-from-scratch/train-text-from-scratch.cpp:1655:35: error: use of ‘auto’ in lambda parameter declaration only available with ‘-std=c++14’ or ‘-std=gnu++14’      
 1655 |     std::sort(begin, end, [&vals](auto a, auto b){                                                                                                                                                           
      |                                   ^~~~                                                                                                                                                                       
/home/mudler/_git/llama.cpp-train/examples/train-text-from-scratch/train-text-from-scratch.cpp:1655:43: error: use of ‘auto’ in lambda parameter declaration only available with ‘-std=c++14’ or ‘-std=gnu++14’      
 1655 |     std::sort(begin, end, [&vals](auto a, auto b){                                                                                                                                                           
      |                                           ^~~~                                                                                                                                                               
/home/mudler/_git/llama.cpp-train/examples/train-text-from-scratch/train
```

xaedes committed:
replace auto parameters in lambda function

xaedes commented:
> JFYI, give it a shot locally, compiling in Linux fails with:

Oh, thanks for pointing that out. Pushed a fix for this to the branch, github seems a bit sluggish today, doesn't show up yet.

https://github.com/ggerganov/llama.cpp/commit/7f172c1070d514e450e002e430957773093572ba

johnson442 commented:
I needed to add `#include <climits>` to train-text-from-scratch.cpp due to:

```
/tmp/mount/train-git/llama.cpp/examples/train-text-from-scratch/train-text-from-scratch.cpp:1512:37: error: ‘INT_MAX’ was not declared in this scope
 1512 |     GGML_ASSERT(size >= 0 && size < INT_MAX);
```
on my system, but then it compiles and runs:
```
root@1ccc33b2ee49:/tmp/mount/train-git/llama.cpp/build/bin# ./train-text-from-scratch -h
usage: ./train-text-from-scratch [options]

options:
  -h, --help                 show this help message and exit
  --vocab-model FNAME        model path from which to load vocab (default 'ggml-vic7b-uncensored-q4_0.bin')
  --train-data FNAME         path from which to load training data (default 'shakespeare.txt')
  --checkpoint-in FNAME      path from which to load training checkpoint (default 'checkpoint.bin')
  --checkpoint-out FNAME     path to save training checkpoint (default 'checkpoint.bin')
  --model-out FNAME          path to save ggml model (default 'ggml-checkpoint-f32.bin')
```

xaedes committed:
add #include <climits>

Sovenok-Hacker commented:
> We may need the new [Sophia Optimizer](https://arxiv.org/abs/2305.14342) for a 2X increase in training speed compared to Adam.

I am trying to use it

xaedes committed:
add explicit cast to fix compile error

"error: non-constant-expression cannot be narrowed from type 'int64_t' (aka 'long long') to 'uint32_t' (aka 'unsigned int') in initializer list [-Wc++11-narrowing]"

xaedes committed:
remove trailing whitespace

KASR commented:
Hi @xaedes, thanks for the example! It builds and runs without errors on my pc. I'm trying to follow the default training example. Can you tell if the following files are the correct inputs:

- https://huggingface.co/TheBloke/Wizard-Vicuna-7B-Uncensored-GGML/tree/main 
- https://github.com/brunoklein99/deep-learning-notes/blob/master/shakespeare.txt

Could you explain what the 'samples after optimization' are? e.g.

![image](https://github.com/ggerganov/llama.cpp/assets/22029863/940446a2-7ad1-4d75-b975-fe56a79eca1b)

xaedes commented:
> Can you tell if the following files are the correct inputs:
> 

You can use any llama model as input, just the vocabulary is needed of it. The one you linked is fine.
As text input you can use whatever you like. Shakespeare was just an example (nanoGPT also had training on shakespeare, this is where it comes from). Your linked shakespeare file is quite good coherent text material.

Yes, weird looking "best samples" with repetitive tokens like "old" indeed occur during early training. 

As you said, the generated samples afterwards are better. 
Especially when generating using main with the exported model, as main has better sampling parameter defaults.

The first time I encountered the repetitive "best samples", I was also worried that it got stuck in some local minimum, but eventually it will overcome it.
"old" just seems to be a frequent token in the training data it saw, so it learned it first. 
I also often encounter other repetitive tokens like "the", "and", "." or newlines on early training.

Let it train more and there will occasionally pop up new words and phrases it learns, until the "best samples" start to look better.

xaedes committed:
add ggml_opt_resume_g which accepts forward and backward cgraphs

xaedes committed:
fix formulas in comments

xaedes committed:
bug fix for ggml_compute_forward_get_rows_back_f32

the result should be set to zero, not to whatever data is in opt0

xaedes committed:
improve training memory usage with scratch buffers

instead of relying on the automatic backward pass, we manually create the graph for the backward pass.
it turns out that all backward pass operations need only temporary memory which can be reused after each layer.

will compute backward pass for ALL model parameters

xaedes committed:
add option to use scratch buffers in training or not

make it configurable because currently training with scratch buffers implies flash attention and optimization over all parameters.

xaedes commented:
> There is still a lot of unnecessary memory overhead.
In the llama eval function similar overhead is avoided using scratch buffers.
Would be good to find a way to do similar with the (automatically generated) backward pass.

Instead of automatically generating the backward pass, I added a function which directly implements the backward pass. This allows to control the memory usage using scratch buffers. 

Turns out that all operations for computing gradients are only temporary necessary. Memory can be reused after each layer.
But the forward activations for all layers are still necessary to remember until the corresponding backward operation.

This greatly reduces memory usage, especially allowing more layers.

xaedes commented:
> Interested in what is the process for creating this function? Is it just an unroll of the generated one or hand-written?

It is hand-written guided by what the automatic backward pass would do. 
The process is pretty straight forward:

First write out the forward process with a unique name for each tensor in topogical order, e.g. post-order.
For each tensor decide if it has a gradient, by looking whether any input tensor to this tensor has a gradient.
I usually write the forward process in comments adding a column for gradients of each tensor. (https://github.com/xaedes/llama.cpp/blob/b58d73ca8c5ea1baf42c24db58746b9e763384af/examples/train-text-from-scratch/train-text-from-scratch.cpp#L1611-L1702)

Then do the backward pass:
Go over the forward pass operations, but in reverse.
Each tensor/operation will add something to the gradients of its input tensors. 
This will be noted in the gradients column of the comment.
If unsure, one can lookup what exactly has to be added in `ggml_compute_backward`.
After all forward pass operations are processed in this manner, we have all necessary gradients noted down.
These can then be directly implemented by calling the corresponding ggml operations.

Writing the forward process in one columns and the backward process gradients in another column right next to it, makes it easy to look for tensors that are used more than once, and decide which tensors are only temporary necessary.

Noting the tensor shapes helps to make sure everything is correct.

I used the same process to derive the backward pass implementations for softmax, cross entropy, etc.

augustoqm commented:
A few more test replications on **Linux/PopOs 22.04 - AMD Ryzen 7**

I did the same setting above - and all went fine. 
```
# wget https://raw.githubusercontent.com/brunoklein99/deep-learning-notes/master/shakespeare.txt
# ./train-text-from-scratch --vocab-model Wizard-Vicuna-7B-Uncensored.ggmlv3.q5_1.bin --mem-compute 12
```

Last output was similar, error decay curve seems promising (as noted, generated text seems better as we go). 

```sh
main: opt->params.adam.sched 0.84000
Example 7, opt iter 100
error_before_opt: 61.881432
error_after_opt:  57.501781
used_mem_before_opt: 842559472 bytes
used_mem_after_opt:  843328080 bytes
Generating 1024 tokens.
 love, thou my love receivest,
I cannot blame thee, for my love thou usest,
But yet be blamed, if thou thy self deceivest
By wilful taste of what thy self refusest.
I do forgive thy robbery gentle thief
Although thou steal thee all my poverty:
And yet love knows it is a greater grief
To bear greater wrong, than hate's known injury.
Lascivious grace, in whom all ill well---
```
I will check with diff data and iterations later. 

BTW Thanks for providing me the super-hero feeling of training on my laptop! :rocket: :superhero: :100:

ggerganov commented:
I will be looking into merging this PR next

ggerganov commented:
Today I noticed the following usage of `padding` to store the `offset`:

https://github.com/ggerganov/llama.cpp/blob/2d7bf110edd8c49209401a16132052cba706ffd0/ggml.c#L5896-L5898

Is it possible to replace these usages with `opt` tensor, as demonstrated here:

https://github.com/ggerganov/llama.cpp/blob/2d7bf110edd8c49209401a16132052cba706ffd0/ggml.c#L5882-L5889

xaedes commented:
> Today I noticed the following usage of `padding` to store the `offset`:

Ohh, this comes from my first implementations of the backward passes. 
Totally forgot to make this right. 
Afaik there is another one in ggml_permute, where I store the axes in the paddings for backward pass.  
Will push a change soon.

xaedes committed:
store view offset and permute axes in opt[0] instead of storing it in padding

use memcpy to store offset, because offset is of type size_t.
when storing it as int32_t offset would have to be smaller than 2^31 which is not necessarily true.

xaedes commented:
> Is it possible to replace these usages with `opt` tensor, as demonstrated here

Done.

niansa commented:
Amazing! Really looking towards the merge :+1:

ggerganov commented:
I tried training a model by running the following command 2 consecutive times:

```bash
./bin/train-text-from-scratch --vocab-model ../models/ggml-vocab.bin --mem-compute 12
```

After the first run, the model was generating something that resembled text, but after the second run it started outputting just `,` or spaces.

Running with thread sanitizer enabled, there seems to be a data race in `ggml_compute_forward_flash_attn_back()`:

<details>

```c
$ ▶ ./bin/train-text-from-scratch --vocab-model ../models/ggml-vocab.bin --mem-compute 12

train-text-from-scratch(2367,0x1f6175e00) malloc: nano zone abandoned due to inability to reserve vm space.
llama.cpp: loading model from ../models/ggml-vocab.bin
llama_model_load_internal: format     = ggjt v1 (pre #1405)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 512
llama_model_load_internal: n_embd     = 4096
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 32
llama_model_load_internal: n_layer    = 32
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: ftype      = 1 (mostly F16)
llama_model_load_internal: n_ff       = 11008
llama_model_load_internal: n_parts    = 1
llama_model_load_internal: model size = 7B
main: tokenize training data
main: number of training tokens: 27584
print_params: n_vocab: 32000
print_params: n_ctx:   128
print_params: n_embd:  256
print_params: n_mult:  256
print_params: n_head:  8
print_params: n_ff:    768
print_params: n_layer: 16
print_params: n_rot:   32
main: number of unique tokens: 3070
main: init model
load_checkpoint: Training iterations: 0.
load_checkpoint: Training samples:    0.
load_checkpoint: Training tokens:     0.
main: opt iter 0
used_mem model+cache: 244461568 bytes
main: begin training
main: opt->params.adam.sched 0.00000
==================
WARNING: ThreadSanitizer: data race (pid=2367)
  Read of size 8 at 0x000c031d0000 by thread T11:
    #0 ggml_compute_forward_flash_attn_back ggml.c:13819 (train-text-from-scratch:arm64+0x10007d44c) (BuildId: d0b59247a937336f9b4e57c596e5a3a832000000200000000100000000000d00)
    #1 ggml_compute_forward ggml.c:14419 (train-text-from-scratch:arm64+0x1000509e4) (BuildId: d0b59247a937336f9b4e57c596e5a3a832000000200000000100000000000d00)
    #2 ggml_graph_compute_thread ggml.c:15519 (train-text-from-scratch:arm64+0x10004fff8) (BuildId: d0b59247a937336f9b4e57c596e5a3a832000000200000000100000000000d00)

  Previous write of size 8 at 0x000c031d0000 by thread T12:
    #0 ggml_compute_forward_flash_attn_back ggml.c:13819 (train-text-from-scratch:arm64+0x10007d468) (BuildId: d0b59247a937336f9b4e57c596e5a3a832000000200000000100000000000d00)
    #1 ggml_compute_forward ggml.c:14419 (train-text-from-scratch:arm64+0x1000509e4) (BuildId: d0b59247a937336f9b4e57c596e5a3a832000000200000000100000000000d00)
    #2 ggml_graph_compute_thread ggml.c:15519 (train-text-from-scratch:arm64+0x10004fff8) (BuildId: d0b59247a937336f9b4e57c596e5a3a832000000200000000100000000000d00)

  Location is heap block of size 8589934592 at 0x000c00010000 allocated by main thread:
    #0 operator new[](unsigned long) <null>:65021472 (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x8ffbc) (BuildId: 981013a59ee23029b2ed90b76951327532000000200000000100000000000b00)
    #1 main train-text-from-scratch.cpp:3104 (train-text-from-scratch:arm64+0x100011a50) (BuildId: d0b59247a937336f9b4e57c596e5a3a832000000200000000100000000000d00)

  Thread T11 (tid=12825379, running) created by main thread at:
    #0 pthread_create <null>:65021472 (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x2fd88) (BuildId: 981013a59ee23029b2ed90b76951327532000000200000000100000000000b00)
    #1 ggml_graph_compute ggml.c:15563 (train-text-from-scratch:arm64+0x10004eeb0) (BuildId: d0b59247a937336f9b4e57c596e5a3a832000000200000000100000000000d00)
    #2 ggml_opt_resume_g ggml.c:17529 (train-text-from-scratch:arm64+0x1000552d4) (BuildId: d0b59247a937336f9b4e57c596e5a3a832000000200000000100000000000d00)
    #3 main train-text-from-scratch.cpp:3198 (train-text-from-scratch:arm64+0x1000120f8) (BuildId: d0b59247a937336f9b4e57c596e5a3a832000000200000000100000000000d00)

  Thread T12 (tid=12825380, running) created by main thread at:
    #0 pthread_create <null>:65021472 (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x2fd88) (BuildId: 981013a59ee23029b2ed90b76951327532000000200000000100000000000b00)
    #1 ggml_graph_compute ggml.c:15563 (train-text-from-scratch:arm64+0x10004eeb0) (BuildId: d0b59247a937336f9b4e57c596e5a3a832000000200000000100000000000d00)
    #2 ggml_opt_resume_g ggml.c:17529 (train-text-from-scratch:arm64+0x1000552d4) (BuildId: d0b59247a937336f9b4e57c596e5a3a832000000200000000100000000000d00)
    #3 main train-text-from-scratch.cpp:3198 (train-text-from-scratch:arm64+0x1000120f8) (BuildId: d0b59247a937336f9b4e57c596e5a3a832000000200000000100000000000d00)

SUMMARY: ThreadSanitizer: data race ggml.c:13819 in ggml_compute_forward_flash_attn_back
==================
```
</details>

ggerganov committed:
minor : fix compile warnings + minor style changes

Entretoize commented:
Sorry for the noob question, but how to build that ?

xaedes commented:
> Running with thread sanitizer enabled, there seems to be a data race in `ggml_compute_forward_flash_attn_back()`:

Ohhh.. thanks! That sounds interesting, will definitely look into that.

xaedes committed:
fix bug in threaded indices calculation of ggml_compute_forward_flash_attn_back_f32

xaedes commented:
@ggerganov 

Pushed a fix for threaded index calculation, I think this probably was it. 

Errors over time [without scratch, with flash] and [without scratch, without flash] are now identical:

Flash
```
Example 7, opt iter 8
error_before_opt: 62.218380
error_after_opt:  62.212128
```

No Flash
```
Example 7, opt iter 8
error_before_opt: 62.218380
error_after_opt:  62.212128
```

But with scratch buffers the errors go down slower:
```
Example 7, opt iter 8
error_before_opt: 62.231804
error_after_opt:  62.228424
```

Will have to take a closer look into that as well.

ggerganov commented on ggml.c:
```diff
@@ -5884,7 +5884,17 @@ struct ggml_tensor * ggml_view_1d(
 
     if (is_node) {
-        memcpy(result->padding, &offset, sizeof(offset));
+        ggml_scratch_save(ctx);
+
+        struct ggml_tensor * b = ggml_new_tensor_1d(ctx, GGML_TYPE_I32, 2);
+
+        GGML_ASSERT(sizeof(offset) <= ggml_nbytes(b));
+
+        memcpy(b->data, &offset, sizeof(offset));
+
+        ggml_scratch_load(ctx);
+
+        result->opt[0] = b;
```
On `master` we now store the "view" offsets even when the tensor is not a `node`:

https://github.com/ggerganov/llama.cpp/blob/fa84c4b3e80199a5683438f062009c031a06c4fa/ggml.c#L5988-L5995

This is used by the Metal inference in `ggml-metal.m` to be able to determine the offsets during forward inference. I think you simply need to remove the `if (is_node)` checks and it should be good

xaedes committed:
store view offset like in master branch

ggerganov commented:
What is a sequence of commands that I can try using to create a model?
Currently, doing the steps that I showed in the previous comment results in non-coherent generation (i.e. lots of `,` and spaces), even though the error keeps going down after every `train-text-from-scratch`. Do I keep running it until it becomes better?

xaedes commented:
> What is a sequence of commands that I can try using to create a model? 

I think the training with scratch buffers might contain some bug, currently searching for it. When training without scratch buffers it gives reasonable output.

Here a verbose command specifing most settings (I used the [shakespeare.txt](https://github.com/brunoklein99/deep-learning-notes/blob/master/shakespeare.txt) from above):

`train-text-from-scratch --vocab-model ggml-model-with-vocab.bin --ctx 64 --embd 256 --head 8 --layer 16 --checkpoint-in chk-shakespeare-256x16.bin --checkpoint-out chk-shakespeare-256x16.bin --model-out ggml-shakespeare-256x16-f32.bin --train-data "shakespeare.txt" -t 6 -b 16 -n 32 --seed 1 --adam-iter 16 --print-details-interval 0 --predict 16 --no-scratch --use-flash`

`main -m ggml-shakespeare-256x16-f32.bin`

Output of main:
<details>

```
And no face in their w heart:
Nor that I I my self thou heart,
O thy worth is thee not all love the to love,
Thy with my heart with loss from me of such them do shall me, and a them not not.
And precious, that love.
In such trust foring thy looks so, and thy self in thee to thy self will,
And disoth physance,,
Dt face.

If the adj receive thee, that I them for thyain,
Who not my love of thy skill,
In be more still more shall all so,
For,
But the best then so, and of my pen.
And is in my love to thee,
O on I have trust thision to keep in the world
As uncious, and will I me,
Bere,
And will all my self I those herion'ed that those receives sorrow,
And-ust for thy self in himce.
Myill
```
</details>

xaedes committed:
bug fix in forward_batch_wo_cache_flash_attn_train

daboe01 commented:
just curious, would it really work out to fine tune llama with approx. 100 MB text on a Apple M2 Max in say one or two weeks?
will this benefit from the recent metal additions?

ggerganov commented:
Thanks @xaedes - the results look much better

> I think the training with scratch buffers might contain some bug, currently searching for it.

Yeah, the scratch buffers are super difficult to use and debug. Need to come up with a better, more automated approach in the future. Anyway, let's see if we can find the issue and then proceed with merging.

xaedes committed:
scratch buffer bug fixes in forward_batch_wo_cache_flash_attn_train

data of permute and reshape is the same as their input.
if we want to preserve the output of permute/reshape, we also need to preserve their inputs.

replace reshape(src0, src1) with reshape_nd calls so that we don't need src1.

replace (temporary) t03 with ggml_repeat(ctx0, layer.attention_norm, t02).
in the future we could also use the new broadcasting ggml_mul to avoid these repeat calls.
for this we need backward pass of broadcasting ggml_mul.

xaedes committed:
remove unnecessary scratch buffer 0

buf 0 is persistent memory, so we can just disable scratch for this by using buf -1

xaedes committed:
avoid creating unnecessary grad tensors

previously we need to create grads for model parameters, so that expand(..) correctly populates cgraph->leafs & cgraph->grads
this wasted memory, because unnecessary grad for each op were automatically created:
the automatically generated grad was unnecessary because we later manually set the grad (e.g. t35->grad = expand(gb, ...) ).
this discarded the automatically generated grad resulting in wasted memory.

improved this by changing expand(..) to not use ggml_build_forward_expand.
expand set cgraph->nodes but not the leafs.
cgraph->leafs & cgraph->grads are set in another pass after the last expand call.

xaedes commented:
> Yeah, the scratch buffers are super difficult to use and debug. Need to come up with a better, more automated approach in the future. Anyway, let's see if we can find the issue and then proceed with merging.

Resolved the scratch buffer issues, now it gives the same error values like the other variants. 

> more automated approach

This would really be helpful. 
Something like:
- first create the tensors in a `noalloc=true` context
- last usage of each tensor determines its required lifetime
- do the magic to assign memory. i think it is related to [register allocation](https://en.wikipedia.org/wiki/Register_allocation), "the process of assigning local automatic variables and expression results to a limited number of processor registers."
- recreate the tensors in the actual context with the final data pointers

Another thought:
The tensors are in post-order, so they are topological sorted.
But actually there is more than one topological sorted order.
Calling ggml_visit_parents in different order results in other topological sorted orders.
The necessary lifetime for each tensor differs between those.
Reordering to reduce the necessary lifetimes would allow more memory reusage.

Igoorx commented:
Hey, thanks for this amazing PR! I was trying to use it with a Japanese training data out of curiosity, but I noticed the examples end up with `�` characters (which aren't in the training data):
<details>
<summary>Example 0 during Training</summary>

![image](https://github.com/ggerganov/llama.cpp/assets/14041768/40fc8be9-a327-42d7-867e-e746bae487be)
</details>
<details>
<summary>Output of the trained model</summary>

![image](https://github.com/ggerganov/llama.cpp/assets/14041768/6eb3ea01-2130-4238-a90c-c836f519c5d7)
</details>

xaedes commented:
> I noticed the examples sometimes contain `�` characters (which aren't in the training data)

Interesting. 
Tokens that are not in training data can be generated, but really should not show up in the printed examples. 
Could be a string encoding issue. I just read the bytes of the training data file and then use the llama tokenizer. Judging on the comment `// split string into utf8 chars` it seems to work on utf-8 encoded strings. 

Try to save the training file with utf8 encoding, maybe that already resolves it?

If not could you give me some training data which results in this `�` characters, so I can reproduce it locally?

Igoorx commented:
>Try to save the training file with utf8 encoding, maybe that already resolves it?

It's already in utf-8, I also tried to save it with BOM but that didn't solve the issue.
Here is the training data: [dataset.txt](https://github.com/ggerganov/llama.cpp/files/11716289/dataset.txt)

Green-Sky commented:
> I noticed the examples sometimes contain � characters (which aren't in the training data)

I think the model pieced together an invalid multibyte(multitoken) utf8 character. Just train longer, and it show happen less and less often. :)

Igoorx commented:
> I think the model pieced together an invalid multibyte(multitoken) utf8 character. Just train longer, and it should happen less and less often. :)

That doesn't seem to be the (only?) reason because the examples aren't generated by the model, they are taken from the training data.

ggerganov commented:
@xaedes 

Shall we merge and continue updating from `master`? I'm worried that I am taking too long to review this

Let's merge / rebase on top of latest `master`. Will run some tests to make sure `ggml` changes didn't break something and even if `train-text-from-scratch` is not 100% complete, we can merge and fix later

xaedes committed:
Merge remote-tracking branch 'upstream/master' into text-from-scratch

# Conflicts:
#	.github/workflows/tidy-post.yml   (was disabled in text-from-scratch branch)
#	examples/CMakeLists.txt           (added train-text-from-scratch example)
#	ggml.c                            (removed the storage of view offset in padding)

xaedes commented:
> Shall we merge and continue updating from `master`? I'm worried that I am taking too long to review this

Sure!

xaedes committed:
print used training seed

xaedes committed:
zero initialize gfbuf and gbbuf

xaedes commented:
> It is already in utf-8, I also tried to save it with BOM but that didn't solve the issue.
> 
> Here is the training data: 

Thanks! Also very nice that you highlighted the corresponding text in your screenshot, makes it much easier to find the passage :)

I am not yet sure what is going on there. We will have to investigate that outside of this PR.

ggerganov commented:
Alright, will do some testing tomorrow to make sure nothing broke and will merge. Thanks again for this cool work. Always great to see your contributions

ggerganov committed:
ci : re-enable workflows + add README for training

daboe01 commented:
(base) hn00033:train-text-from-scratch daboe01$ ../../train-text-from-scratch \
>         --vocab-model ../models/ggml-vocab.bin \
>         --ctx 64 --embd 256 --head 8 --layer 16 \
>         --checkpoint-in  chk-shakespeare-256x16.bin \
>         --checkpoint-out chk-shakespeare-256x16.bin \
>         --model-out ggml-shakespeare-256x16-f32.bin \
>         --train-data "shakespeare.txt" \
>         -t 6 -b 16 -n 32 --seed 1 --adam-iter 16 \
>         --print-details-interval 0 --predict 16 --use-flash
main: seed: 1
llama.cpp: loading model from ../models/ggml-vocab.bin
Segmentation fault: 11

daboe01 commented:
any plans to support fine-tuning?

xaedes commented:
> llama.cpp: loading model from ../models/ggml-vocab.bin
> Segmentation fault: 11

Does this model file exist? 
The model with vocabulary is loaded by llama.cpp code, and I have seen it crash ungracefully when the model file does not exist.
At this point (https://github.com/ggerganov/llama.cpp/blob/92549202659fc23ba9fec5e688227d0da9b06b40/llama.cpp#L460) file.fp is NULL when the file does not exist and trying to read from it crashes, might need to add a check for this case.

xaedes commented:
@daboe01 

> any plans to support fine-tuning?

@matthiasgeihs
> how is it going with Lora fine-tuning? Would be so cool to have this. Thanks for the great work!

This is absolutely planned! 

Before touching the big models I wanted to further improve the memory usage, which will also result in better performance when fine-tuning. The implementations necessary for training text from scratch already improved memory usage by several orders of magnitude compared to the baby-llama training example.

Currently the memory usage is dominated roughly 50/50 by the optimizer itself and by the backward pass.

When scaling the model size up, the optimizer starts to eat a lot of memory. It requires seven times as much memory as the model itself. This can easily be reduced to only require 2-3 times the model size.

During backward pass all forward layer activations need to be stored, resulting in linearly increasing memory requirements w.r.t to the number of layers. With [gradient checkpointing](https://github.com/cybertronai/gradient-checkpointing) this can be reduced to sqrt(number of layers) while only requiring one additional forward pass. For 32 layers this would be an improvement of 5-6x.
Since the backward pass is now manually implemented, adding gradient checkpointing is not very hard.

Currently F32 is used for training everywhere, but F16 or mixed precision should be enough - enabling up to 2x memory improvement.

Altogether this could (optimistically estimated) reduce the memory usage by ~8x.

That said, I might begin experimenting with LoRa fine-tuning before the memory improvements are done. 
Well LoRa is in itself a kind of memory usage improvement..^^

Green-Sky commented:
> ```
>  hn00033:train-text-from-scratch daboe01$ ../../train-text-from-scratch \
> >         --vocab-model ../models/ggml-vocab.bin \
> >         --ctx 64 --embd 256 --head 8 --layer 16 \
> >         --checkpoint-in  chk-shakespeare-256x16.bin \
> >         --checkpoint-out chk-shakespeare-256x16.bin \
> >         --model-out ggml-shakespeare-256x16-f32.bin \
> >         --train-data "shakespeare.txt" \
> >         -t 6 -b 16 -n 32 --seed 1 --adam-iter 16 \
> >         --print-details-interval 0 --predict 16 --use-flash
> main: seed: 1
> llama.cpp: loading model from ../models/ggml-vocab.bin
> Segmentation fault: 11
> ```

try reducing batch size, larger batches segfault for me too.

daboe01 commented:
> @daboe01
> 
> > llama.cpp: loading model from ../models/ggml-vocab.bin
> > Segmentation fault: 11
> 
> Does this model file exist? The model with vocabulary is loaded by llama.cpp code, and I have seen it crash ungracefully when the model file does not exist. At this point (
> 
> https://github.com/ggerganov/llama.cpp/blob/92549202659fc23ba9fec5e688227d0da9b06b40/llama.cpp#L460
> 
> ) file.fp is NULL when the file does not exist and trying to read from it crashes, might need to add a check for this case.

you are right!
this fixed the segfault.
see #1850

matthiasgeihs commented:
> Well LoRa is in itself a kind of memory usage improvement..^^

exactly ;) @xaedes

fpena06 commented:
Hello,

I'm trying to test this on a Mac M2. I did a git pull, then make. But it's not compiling train-text-from-scratch, I do see the train-text-from-scratch folder in examples. Any suggestions?

Thanks

daboe01 commented:
> Hello,
> 
> I'm trying to test this on a Mac M2. I did a git pull, then make. But it's not compiling train-text-from-scratch, I do see the train-text-from-scratch folder in examples. Any suggestions?
> 
> ```
> (llamacpp) ~/Downloads/llama.cpp$ make
> I llama.cpp build info:
> I UNAME_S:  Darwin
> I UNAME_P:  arm
> I UNAME_M:  arm64
> I CFLAGS:   -I.              -O3 -std=c11   -fPIC -DNDEBUG -Wall -Wextra -Wpedantic -Wcast-qual -Wdouble-promotion -Wshadow -Wstrict-prototypes -Wpointer-arith -pthread -DGGML_USE_K_QUANTS -DGGML_USE_ACCELERATE
> I CXXFLAGS: -I. -I./examples -O3 -std=c++11 -fPIC -DNDEBUG -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wno-multichar -pthread -DGGML_USE_K_QUANTS
> I LDFLAGS:   -framework Accelerate
> I CC:       Apple clang version 14.0.3 (clang-1403.0.22.14.1)
> I CXX:      Apple clang version 14.0.3 (clang-1403.0.22.14.1)
> 
> cc  -I.              -O3 -std=c11   -fPIC -DNDEBUG -Wall -Wextra -Wpedantic -Wcast-qual -Wdouble-promotion -Wshadow -Wstrict-prototypes -Wpointer-arith -pthread -DGGML_USE_K_QUANTS -DGGML_USE_ACCELERATE   -c ggml.c -o ggml.o
> c++ -I. -I./examples -O3 -std=c++11 -fPIC -DNDEBUG -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wno-multichar -pthread -DGGML_USE_K_QUANTS -c llama.cpp -o llama.o
> llama.cpp:877:38: warning: unused parameter 'n_gpu_layers' [-Wunused-parameter]
>                                int   n_gpu_layers) {
>                                      ^
> 1 warning generated.
> c++ -I. -I./examples -O3 -std=c++11 -fPIC -DNDEBUG -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wno-multichar -pthread -DGGML_USE_K_QUANTS -c examples/common.cpp -o common.o
> cc -I.              -O3 -std=c11   -fPIC -DNDEBUG -Wall -Wextra -Wpedantic -Wcast-qual -Wdouble-promotion -Wshadow -Wstrict-prototypes -Wpointer-arith -pthread -DGGML_USE_K_QUANTS -DGGML_USE_ACCELERATE   -c -o k_quants.o k_quants.c
> c++ -I. -I./examples -O3 -std=c++11 -fPIC -DNDEBUG -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wno-multichar -pthread -DGGML_USE_K_QUANTS examples/main/main.cpp ggml.o llama.o common.o k_quants.o -o main  -framework Accelerate
> 
> ====  Run ./main -h for help.  ====
> 
> c++ -I. -I./examples -O3 -std=c++11 -fPIC -DNDEBUG -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wno-multichar -pthread -DGGML_USE_K_QUANTS examples/quantize/quantize.cpp ggml.o llama.o k_quants.o -o quantize  -framework Accelerate
> c++ -I. -I./examples -O3 -std=c++11 -fPIC -DNDEBUG -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wno-multichar -pthread -DGGML_USE_K_QUANTS examples/quantize-stats/quantize-stats.cpp ggml.o llama.o k_quants.o -o quantize-stats  -framework Accelerate
> c++ -I. -I./examples -O3 -std=c++11 -fPIC -DNDEBUG -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wno-multichar -pthread -DGGML_USE_K_QUANTS examples/perplexity/perplexity.cpp ggml.o llama.o common.o k_quants.o -o perplexity  -framework Accelerate
> c++ -I. -I./examples -O3 -std=c++11 -fPIC -DNDEBUG -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wno-multichar -pthread -DGGML_USE_K_QUANTS examples/embedding/embedding.cpp ggml.o llama.o common.o k_quants.o -o embedding  -framework Accelerate
> c++ -I. -I./examples -O3 -std=c++11 -fPIC -DNDEBUG -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wno-multichar -pthread -DGGML_USE_K_QUANTS pocs/vdot/vdot.cpp ggml.o k_quants.o -o vdot  -framework Accelerate
> ```
> 
> Thanks

@fpena06 take a look at PR #1850

matthiasgeihs commented:
@fpena06 I ran into the same problem. However, it builds if you follow the cmake build steps...

https://github.com/ggerganov/llama.cpp/blob/254a7a7a5ff4c874ff8488f1f5cbdd7e9c89d682/README.md?plain=1#L226-L233

dranger003 commented:
Running on Windows 11 built with MSVC there is nullptr for `tensor->src0` so the program is crashing.

https://github.com/ggerganov/llama.cpp/blob/d411968e990c37f51328849c96a743dd78f3c3dd/ggml-cuda.cu#L2555

```
bin\train-text-from-scratch.exe
        --vocab-model ..\..\models\ggml-vocab.bin
        --ctx 64 --embd 256 --head 8 --layer 16
        --checkpoint-in chk-shakespeare-256x16.bin
        --checkpoint-out chk-shakespeare-256x16.bin
        --model-out ggml-shakespeare-256x16-f32.bin
        --train-data "shakespeare.txt"
        -t 6 -b 16 -n 32 --seed 1 --adam-iter 16
        --print-details-interval 0 --predict 16 --use-flash
```

Call stack:
```
>	train-text-from-scratch.exe!ggml_cuda_compute_forward(ggml_compute_params * params, ggml_tensor * tensor) Line 2554	C++
 	train-text-from-scratch.exe!ggml_compute_forward(ggml_compute_params * params, ggml_tensor * tensor) Line 14314	C
 	train-text-from-scratch.exe!ggml_graph_compute(ggml_context * ctx, ggml_cgraph * cgraph) Line 16018	C
 	train-text-from-scratch.exe!main(int argc, char * * argv) Line 3262	C++
```

Output:
```
main: seed: 1
llama.cpp: loading model from ..\..\models\ggml-vocab.bin
llama_model_load_internal: format     = ggjt v1 (pre #1405)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 512
llama_model_load_internal: n_embd     = 4096
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 32
llama_model_load_internal: n_layer    = 32
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: ftype      = 1 (mostly F16)
llama_model_load_internal: n_ff       = 11008
llama_model_load_internal: n_parts    = 1
llama_model_load_internal: model size = 7B
main: tokenize training data
main: number of training tokens: 27584
print_params: n_vocab: 32000
print_params: n_ctx:   64
print_params: n_embd:  256
print_params: n_mult:  256
print_params: n_head:  8
print_params: n_ff:    768
print_params: n_layer: 16
print_params: n_rot:   32
main: number of unique tokens: 3070
ggml_init_cublas: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 3070 Ti Laptop GPU
main: init model
load_checkpoint: Training iterations: 0.
load_checkpoint: Training samples:    0.
load_checkpoint: Training tokens:     0.
main: opt iter 0
used_mem model+cache: 242364416 bytes
main: begin training
```

daboe01 commented:
is it possible to use a custom vocabulary for a model created with train-from-scratch?

Green-Sky commented:
The first thing you choose, when you start train-from-scratch, is a vocabulary. so yes, you can probably use the convert.py to convert a vocab.json to a ggml-vocab.bin.

daboe01 commented:
@Green-Sky will such a model with custom vocab work out of the box with ./main at inference time?

Green-Sky commented:
Yes. There are muliple models and finetunes outhere with different/modified vocabulary. (eg open_llama has a totally different vocab)

aseok commented:
Is it possible to pass memory related values below defaults? seems example code with default values aborts because of default memory values despite enough memory being available, e.g for ctx 64 (other parameters passed as in example page) it requires 2gb of memory for completion but crashes in 8gb memory system.

xaedes commented:
> Is it possible to pass memory related values below defaults?

There are command line options to specify the buffer sizes (in whole gigabytes).

```
  --mem-model N              Memory to allocate for model and cache in gigabytes. (default 2)
  --mem-compute N            Memory to allocate for compute in gigabytes. (default 24)
  --mem-compute0 N           Memory to allocate for compute in gigabytes. (default 8)
  --mem-compute1 N           Memory to allocate for compute in gigabytes. (default 2)
```

I must admit the default values are probably not the sanest in the world. 
Could be reduced to lower values, especially the `--mem-compute` default of 24 gb is probably too much for small models.
`--mem-compute0` and `--mem-compute1` for small models does not require even one gigabyte, so 1 gb would be enough.

xaedes commented:
>LoRa fine-tuning would be absolutely amazing!

I am nearly done with the before mentioned memory improvements. 
With the implemented optimizations it should already be possible to train from scratch with model size of llama 6b when you have 128gb RAM. I am developing on a 64gb system tho..

`--ctx 256 --embd 1024 --head 8 --layer 32` requires ~11gb
`--ctx 256 --embd 2048 --head 8 --layer 32` requires ~35gb
`--ctx 256 --embd 4096 --head 8 --layer 8` requires ~32gb

Especially gradient checkpointing helped a lot!

Mixed precision training requires F32 of the model parameters, which currently dominates the required memory.
So I think I will skip mixed precision training, leaving it for later, and make a PR with gradient checkpointing and adam optimizer improvements and then continue with working on LORA finetuning.

LORA on quantized parameters should then work out-of-the-box because we need to add the LORA generated matrices to the (quantizied) model parameters which can already be done due to `ggml_compute_forward_add_q_f32`.

xaedes commented:
> Btw, I was recently running ADAM on some of the unit tests in the `ggml` repo and it now fails to optimize 

Thanks for the info! I will look what is going on there.

Igoorx commented:
It might be worthwhile to take a look at this new paper: https://arxiv.org/pdf/2305.14342.pdf
![capture](https://github.com/ggerganov/llama.cpp/assets/14041768/811f93a7-df63-49b5-ab6e-a8f01bdf8da0)

xaedes commented:
> It might be worthwhile to take a look at this new paper: https://arxiv.org/pdf/2305.14342.pdf 

In principle the Sophia-G optimizer does not look _too_ hard to implement.
Only problem is, that it requires access to the (cross entropy) loss function _and_ the logits.
Currently the optimizers accept as input only the final loss function, so it might get a bit ugly to integrate in the existing API.

alasdairforsythe commented:
@xaedes

Since we last spoke I've spent 16 hours a day, 7 days per week working on a way I can contribute to LLM development using my knowledge and experience. The result is a new tokenizer: [TokenMonster](https://github.com/alasdairforsythe/tokenmonster)

I ported [LLaMa Tokenizer](https://bot.co/tokenmonster/?a=llama&b=englishcode-32000-balanced-v1&text=alice), which it tokenizes [more efficiently](https://bot.co/tokenmonster/benchmark.html?a=llama%20huggingface&b=llama%20tokenmonster&c=englishcode-32000-balanced-v1) using an ungreedy tokenization algorithm, and it tokenizes at speeds up to [12.5MB per second](https://github.com/alasdairforsythe/tokenmonster/tree/main/benchmark#readme). The vocabulary and IDs are exactly the same, assuming the LLaMa Tokenizer on Hugging Face is correct (which is where I took it from.)

TokenMonster trained vocabularies are optimal for the dataset and tokenization algorithm. For pretraining model, by using a TokenMonster vocabulary, you can reduce the vocabulary size to [8000 or less and still achieve 4 characters / token](https://bot.co/tokenmonster/benchmark.html?a=fiction-4096-balanced-v1&b=englishcode-8000-clean-v1&c=englishcode-16000-unfiltered-v1). Or you can [keep it at 32000 and get 38% more efficient tokenization](https://bot.co/tokenmonster/?a=llama&b=englishcode-32000-balanced-v1&text=alice).

I'd recommend [englishcode-8000-consistent](https://bot.co/tokenmonster/?a=englishcode-8000-consistent-v1&b=llama&text=alice) for baby llama. It a quarter reduction in model size, has comparable performance with LLaMa Tokenizer, and the "consistent" optimization mode helps smaller models more easily learn the grammar (in theory.) "englishcode" means it was trained on [English](https://github.com/alasdairforsythe/tokenmonster#englishcode) and [code in 30 languages](https://github.com/alasdairforsythe/tokenmonster#code).

There's no C++ yet, it's in Python, Go & Javascript. A C++ version is on the cards, but not in the near future as it relies upon a lot of very highly optimized code I'd previously written in Go, and I'd have to port it all over.

I remember reading that there's a Python frontend for training your baby llama... I can't find it now though. Can you point me to that? If you have the training working now from C++ and have an interface into it from Python, I'm willing to expand that Python frontend out into something user-friendly people can start training small models with. Please point me in the right direction as I can't find any documentation or usage for baby llama.

daboe01 commented:
#2246

DamascusGit commented:
> > LoRa fine-tuning would be absolutely amazing!
> 
> I am nearly done with the before mentioned memory improvements. With the implemented optimizations it should already be possible to train from scratch with model size of llama 6b when you have 128gb RAM. I am developing on a 64gb system tho..
> 
> `--ctx 256 --embd 1024 --head 8 --layer 32` requires ~11gb `--ctx 256 --embd 2048 --head 8 --layer 32` requires ~35gb `--ctx 256 --embd 4096 --head 8 --layer 8` requires ~32gb
> 
> Especially gradient checkpointing helped a lot!
> 
> Mixed precision training requires F32 of the model parameters, which currently dominates the required memory. So I think I will skip mixed precision training, leaving it for later, and make a PR with gradient checkpointing and adam optimizer improvements and then continue with working on LORA finetuning.
> 
> LORA on quantized parameters should then work out-of-the-box because we need to add the LORA generated matrices to the (quantized) model parameters which can already be done due to `ggml_compute_forward_add_q_f32`.

so we can QLoRA on 4-bit models with ggml on metal?

Sovenok-Hacker commented:
> @xaedes
> 
> Since we last spoke I've spent 16 hours a day, 7 days per week working on a way I can contribute to LLM development using my knowledge and experience. The result is a new tokenizer: [TokenMonster](https://github.com/alasdairforsythe/tokenmonster)
> 
> I ported [LLaMa Tokenizer](https://bot.co/tokenmonster/?a=llama&b=englishcode-32000-balanced-v1&text=alice), which it tokenizes [more efficiently](https://bot.co/tokenmonster/benchmark.html?a=llama%20huggingface&b=llama%20tokenmonster&c=englishcode-32000-balanced-v1) using an ungreedy tokenization algorithm, and it tokenizes at speeds up to [12.5MB per second](https://github.com/alasdairforsythe/tokenmonster/tree/main/benchmark#readme). The vocabulary and IDs are exactly the same, assuming the LLaMa Tokenizer on Hugging Face is correct (which is where I took it from.)
> 
> TokenMonster trained vocabularies are optimal for the dataset and tokenization algorithm. For pretraining model, by using a TokenMonster vocabulary, you can reduce the vocabulary size to [8000 or less and still achieve 4 characters / token](https://bot.co/tokenmonster/benchmark.html?a=fiction-4096-balanced-v1&b=englishcode-8000-clean-v1&c=englishcode-16000-unfiltered-v1). Or you can [keep it at 32000 and get 38% more efficient tokenization](https://bot.co/tokenmonster/?a=llama&b=englishcode-32000-balanced-v1&text=alice).
> 
> I'd recommend [englishcode-8000-consistent](https://bot.co/tokenmonster/?a=englishcode-8000-consistent-v1&b=llama&text=alice) for baby llama. It a quarter reduction in model size, has comparable performance with LLaMa Tokenizer, and the "consistent" optimization mode helps smaller models more easily learn the grammar (in theory.) "englishcode" means it was trained on [English](https://github.com/alasdairforsythe/tokenmonster#englishcode) and [code in 30 languages](https://github.com/alasdairforsythe/tokenmonster#code).
> 
> There's no C++ yet, it's in Python, Go & Javascript. A C++ version is on the cards, but not in the near future as it relies upon a lot of very highly optimized code I'd previously written in Go, and I'd have to port it all over.
> 
> I remember reading that there's a Python frontend for training your baby llama... I can't find it now though. Can you point me to that? If you have the training working now from C++ and have an interface into it from Python, I'm willing to expand that Python frontend out into something user-friendly people can start training small models with. Please point me in the right direction as I can't find any documentation or usage for baby llama.

I am interested in your tokenizer, trained a vocab, it works great. It will be good to integrate it to HF Tokenizers and LLaMA.cpp

Sovenok-Hacker commented:
> > It might be worthwhile to take a look at this new paper: https://arxiv.org/pdf/2305.14342.pdf
> 
> In principle the Sophia-G optimizer does not look _too_ hard to implement. Only problem is, that it requires access to the (cross entropy) loss function _and_ the logits outputting function. Currently the optimizers accept as input only the final loss function, so it might get a bit ugly to integrate in the existing API.
> 
> Edit: Just passing the logits function instead of the loss function to the optimizer and then computing the cross entropy loss inside Sophia-G should work. But even then it still needs to be passed the target probabilities, hm...

It will be good to do it
